crates/sui-indexer-alt-graphql/src/api/types/unchanged_shared_object.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_graphql::{Object, SimpleObject, Union};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    digests::ObjectDigest,
+    effects::UnchangedSharedKind as NativeUnchangedSharedKind,
+};
+
+use crate::{
+    api::scalars::{sui_address::SuiAddress, uint53::UInt53},
+    scope::Scope,
+};
+
+use super::{address::Address, object::Object};
+
+/// Details pertaining to shared objects that are referenced by but not changed by a transaction.
+#[derive(Union)]
+pub(crate) enum UnchangedSharedObject {
+    Read(SharedObjectRead),
+    MutateStreamEnded(SharedObjectMutateStreamEnded),
+    ReadStreamEnded(SharedObjectReadStreamEnded),
+    Cancelled(SharedObjectCancelled),
+    PerEpochConfig(SharedObjectPerEpochConfig),
+}
+
+pub(crate) struct SharedObjectRead {
+    scope: Scope,
+    object_id: ObjectID,
+    version: SequenceNumber,
+    digest: ObjectDigest,
+}
+
+#[Object]
+impl SharedObjectRead {
+    /// The shared object that was accessed read-only.
+    async fn object(&self) -> Option<Object> {
+        let address = Address::with_address(self.scope.clone(), self.object_id.into());
+        Some(Object::with_ref(address, self.version, self.digest))
+    }
+}
+
+/// Access to a shared object whose mutable consensus stream has ended.
```

Comment:
This input kind corresponds to a transaction that wanted to mutate a consensus-managed object but couldn't because it became not-consensus-managed before then (either it was deleted, or turned into an owned object).

We should also keep with the nomenclature in the `Native` version of this type (`Consensus` rather than `Shared`) because this will be used for handling party objects and consensus-managed address-owned objects as well.